examples/question-answering/run_qa.py

```diff
@@ -360,13 +374,17 @@ def prepare_train_features(examples):
         return tokenized_examples
 
     if training_args.do_train:
+        if "train" not in datasets:
+            raise ValueError("--do_train requires a train dataset")
         train_dataset = datasets["train"].map(
             prepare_train_features,
             batched=True,
             num_proc=data_args.preprocessing_num_workers,
             remove_columns=column_names,
             load_from_cache_file=not data_args.overwrite_cache,
         )
+        if data_args.max_train_samples is not None:
+            train_dataset = train_dataset.select(range(data_args.max_train_samples))
```

Comment:
We should still do this before the map: the map adds samples but do not reduce their numbers. So in this example, we can speed up preprocessing by doing the `train_dataset = train_dataset.select(range(data_args.max_train_samples))
` before the map so we preprocess at most `max_train_samples` examples, and then once more after the map to make sure we have the right number of examples.